source/fuzz/transformation_propagate_instruction_up.cpp

```diff
@@ -0,0 +1,393 @@
+// Copyright (c) 2020 Vasyl Teliman
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_propagate_instruction_up.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+namespace {
+
+std::unordered_map<uint32_t, std::unordered_map<uint32_t, uint32_t>>
+ComputeMappingFromOpPhiToResultId(opt::IRContext* ir_context,
+                                  opt::Instruction* inst) {
+  std::unordered_map<uint32_t, std::unordered_map<uint32_t, uint32_t>> result;
+
+  const auto* inst_block = ir_context->get_instr_block(inst);
+
+  for (uint32_t i = 0; i < inst->NumInOperands(); ++i) {
+    const auto& operand = inst->GetInOperand(i);
+    if (operand.type != SPV_OPERAND_TYPE_ID) {
+      continue;
+    }
+
+    auto* dependency = ir_context->get_def_use_mgr()->GetDef(operand.words[0]);
+    assert(dependency && "|inst| depends on invalid id");
+
+    if (ir_context->get_instr_block(dependency) == inst_block &&
+        dependency->opcode() == SpvOpPhi &&
+        !result.count(dependency->result_id())) {
+      std::unordered_map<uint32_t, uint32_t> label_id_to_result_id;
+
+      for (uint32_t j = 1; j < dependency->NumInOperands(); j += 2) {
+        label_id_to_result_id[dependency->GetSingleWordInOperand(j)] =
+            dependency->GetSingleWordInOperand(j - 1);
+      }
+
+      result[dependency->result_id()] = std::move(label_id_to_result_id);
+    }
+  }
+
+  return result;
+}
+
+}  // namespace
+
+TransformationPropagateInstructionUp::TransformationPropagateInstructionUp(
+    const protobufs::TransformationPropagateInstructionUp& message)
+    : message_(message) {}
+
+TransformationPropagateInstructionUp::TransformationPropagateInstructionUp(
+    uint32_t block_id,
+    const std::unordered_map<uint32_t, uint32_t>& predecessor_id_to_fresh_id) {
+  message_.set_block_id(block_id);
+  message_.mutable_predecessor_id_to_fresh_id()->insert(
+      predecessor_id_to_fresh_id.begin(), predecessor_id_to_fresh_id.end());
+}
+
+bool TransformationPropagateInstructionUp::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // Check that we can apply this transformation to the |block_id|.
+  if (!IsApplicableToTheBlock(ir_context, message_.block_id())) {
```

Comment:
I would change the name to `IsApplicableToBlock` - it's typical to drop articles when writing variable and function names (though there are no hard rules.)

Reply:
Fixed in c74f0e3.